compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -370,15 +500,31 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
 object InstrumentCoverage:
   val name: String = "instrumentCoverage"
   val description: String = "instrument code for coverage checking"
-
-  /** Extractor object for trees produced by `insertInvokeCall`. */
-  object InstrumentedBlock:
-    private def isInvokedCall(app: Apply)(using Context): Boolean =
-      app.span.isSynthetic && app.symbol == defn.InvokedMethodRef.symbol
-
-    def unapply(t: Tree)(using Context): Option[(Apply, Tree)] =
-      t match
-        case Block((app: Apply) :: Nil, expr) if isInvokedCall(app) =>
-          Some((app, expr))
-        case _ =>
-          None
+  val ExcludeMethodFlags: FlagSet = Synthetic | Artifact | Erased
+
+  /**
+   * An instrumented Tree, in 3 parts.
+   * @param pre preparation code, e.g. lifted arguments. May be empty.
+   * @param invokeCall call to Invoker.invoked(dir, id), or an empty tree.
+   * @param expr the instrumented expression, executed just after the invokeCall
+   */
+  case class InstrumentedParts(pre: List[Tree], invokeCall: Apply | Thicket, expr: Tree):
```

Comment:
If we want to be precise, could this be:
```suggestion
  case class InstrumentedParts(pre: List[Tree], invokeCall: Apply | EmptyTree.type, expr: Tree):
```